app/lib/analyzer/backend.dart

```diff
@@ -0,0 +1,158 @@
+  /// be returned.
+  Future<Analysis> getAnalysis(String package,
+      [String version, int analysis]) async {
+    final Key packageKey = db.emptyKey.append(PackageAnalysis, id: package);
+
+    if (version == null) {
+      final list = await db.lookup([packageKey]);
+      final PackageAnalysis pa = list[0];
+      if (pa == null) return null;
+      version = pa.latestVersion;
+      analysis = pa.latestAnalysis;
+    }
+
+    // version was set
+    final Key versionKey =
+        packageKey.append(PackageVersionAnalysis, id: version);
+
+    if (analysis == null) {
+      final list = await db.lookup([versionKey]);
+      final PackageVersionAnalysis pva = list[0];
+      if (pva == null) return null;
+      analysis = pva.latestAnalysis;
+    }
+
+    // analysis was set
+    final Key analysisKey = versionKey.append(Analysis, id: analysis);
+    final List list = await db.lookup([analysisKey]);
+    return list[0];
+  }
+
+  /// Create a new [Analysis] object with the current timestamp.
+  Analysis prepareAnalysis(String package, String version) {
+    final DateTime now = new DateTime.now().toUtc();
+    final Key versionKey = db.emptyKey
+        .append(PackageAnalysis, id: package)
+        .append(PackageVersionAnalysis, id: version);
+    return new Analysis()
+      ..parentKey = versionKey
+      ..packageName = package
+      ..packageVersion = version
+      ..timestamp = now;
+  }
+
+  /// Stores the analysis, and either creates or updates its parent
+  /// [PackageAnalysis] and [PackageVersionAnalysis] rollups records.
+  Future storeAnalysis(Analysis analysis) async {
+    assert(analysis.id == null);
+    await db.commit(inserts: [analysis]);
+    assert(analysis.id != null);
+
+    // update package and version too
+    await db.withTransaction((Transaction tx) async {
+      final Key packageKey =
+          db.emptyKey.append(PackageAnalysis, id: analysis.packageName);
+      final Key packageVersionKey = packageKey.append(PackageVersionAnalysis,
+          id: analysis.packageVersion);
+      final List parents = await tx.lookup([packageKey, packageVersionKey]);
+      final PackageAnalysis package = parents[0];
+      final PackageVersionAnalysis version = parents[1];
+
+      final List<Model> inserts = [];
+      if (version == null) {
+        inserts.add(_createVersion(analysis));
+      } else {
+        if (_updateVersionLatest(version, analysis)) {
+          inserts.add(version);
+        }
+      }
+
+      // If there is no change in the version, there won't be a change in the
+      // package either.
+      //
+      // Note: [PackageAnalysis] doesn't store analyzerVersion, it relies on
+      // [PackageVersionAnalysis] to always store the latest.
+      if (inserts.isEmpty) {
+        await tx.rollback();
+        return;
+      }
+
+      if (package == null) {
+        inserts.add(_createPackage(analysis));
+      } else {
+        if (_updatePackageLatest(package, analysis)) {
```

Comment:
This is updating `PackageAnalysis` purely based on the new `Analysis` irrespective of the semver of the pub version (e.g. `PackageAnalysis.latestAnalysis` is for 1.0.0 with pana 0 and analysis is 0.1.0 with pana 1, so here we don't want to update `latestAnalysis` (at least that is what I thought)).

We can avoid this issue by letting `PackageAnalysis` always point to the latest stable semver `PackageVersionAnalysis` and let this one point to the latest pana version of `Analysis`.

This was I mentioned in the last 2 reviews, that the `Key` in `PackageAnalysis` points to `PackageVersionAnalysis` (and not to an `Analysis` directly).